# Reference: 

# actual data source: https://stats.espncricinfo.com/ci/content/records/93276.html

**import required libraries**

In [140]:
import pandas as pd
import numpy as np

In [141]:
pip install openpyxl

Note: you may need to restart the kernel to use updated packages.


In [142]:
#for reading a excel file 
df = pd.read_excel("test_cricket.xlsx" , sheet_name='wickets')

# Naming the DataFrame - df
# Reading the .xlsx file using pandas: pd.read_excel
# Reading the sheet named 'wickets'

display(df.head(11))

#df.head(11) is for showing first 10 rows of the data frame.

Player       Span  Mat  Inns  Balls   Runs  Wkts  \
0   M Muralitharan (ICC/SL)  1992-2010  133   230  44039  18180   800   
1            SK Warne (AUS)  1992-2007  145   273  40705  17995   708   
2          A Kumble (INDIA)  1990-2008  132   236  40850  18355   619   
3         JM Anderson (ENG)  2003-2021  162   301  34791  16457   617   
4          GD McGrath (AUS)  1993-2007  124   243  29248  12186   563   
5           SCJ Broad (ENG)  2007-2021  148   272  29713  14502   523   
6             CA Walsh (WI)  1984-2001  132   242  30019  12688   519   
7             DW Steyn (SA)  2004-2019   93   171  18608  10077   439   
8       N Kapil Dev (INDIA)  1978-1994  131   227  27740  12867   434   
9         HMRKB Herath (SL)  1999-2018   93   170  25993  12157   433   
10       Sir RJ Hadlee (NZ)  1973-1990   86   150  21918   9611   431   

                    BBI                  BBM    Ave  Econ    SR   5  10  
0   1951-09-01 00:00:00               16/220  22.72  2.47  55.0  67  22  
1   1971-08-01 00:00:00               12/128  25.41  2.65  57.4  37  10  
2   1974-10-01 00:00:00               14/149  29.65  2.69  65.9  35   8  
3   1942-07-01 00:00:00  1971-11-01 00:00:00  26.67  2.83  56.3  30   3  
4   2021-08-24 00:00:00  2021-10-27 00:00:00  21.64  2.49  51.9  29   3  
5   2021-08-15 00:00:00               11/121  27.72  2.92  56.8  18   3  
6   1937-07-01 00:00:00                13/55  24.44  2.53  57.8  22   3  
7   1951-07-01 00:00:00  1960-11-01 00:00:00  22.95  3.24  42.3  26   5  
8   1983-09-01 00:00:00               11/146  29.64  2.78  63.9  23   2  
9                 9/127               14/184  28.07  2.80  60.0  34   9  
10  1952-09-01 00:00:00               15/123  22.29  2.63  50.8  36   9

**Observations:**

In this data set there are 14 columns .

 * player = player name
 

 * span = the staring year and the ending year of a player to play/ career duration
 

 * Mat = number of matches



* inns = number of innings



* balls = total number of bowling. a bowler delivers the ball from his or her end of the pitch towards the batsman



* Runs =  The number of runs conceded


* Wkts= total number wicket taken of a bowler.



* BBI	 = BBI stands for Best Bowling in Innings and only gives the score for one innings. 



* BBM =  BBM stands for Best Bowling in Match and gives the combined score over 2 or more innings in one match. 



* average = bowling average is the number of runs they have conceded per wicket taken.



* Econ = Economy rate.  a player's economy rate is the average number of runs they have conceded per over bowled.



* SR	= strike rate. Bowling Strike Rate is a measurement of a bowler’s average number of balls bowled for every wicket 
taken. 



* 5	= number of times the player took 5 wickets in a match



 * 10 = number of times the player took 10 wickets in a match

In [143]:
# number of rows
print("number of rows = ", df.shape[0])

# number of columns
print("number of columns = ", df.shape[1] )

number of rows =  79
number of columns =  14


**Observations:**

The DataFrame contains 79 rows (players) and 14 columns (attributes).

In [144]:
df.describe()

#checking the data statistic 
#The describe () function is used to generate descriptive statistics 

Mat        Inns         Balls          Runs        Wkts  \
count   79.000000   79.000000     79.000000     79.000000   79.000000   
mean    80.101266  144.797468  18630.303797   8595.506329  317.101266   
std     28.537692   51.042310   7190.036515   3080.256645  121.731587   
min     37.000000   67.000000   8785.000000   4846.000000  200.000000   
25%     60.500000  110.000000  13580.000000   6456.500000  229.000000   
50%     71.000000  129.000000  16498.000000   7742.000000  266.000000   
75%     93.000000  169.000000  21742.500000   9756.000000  374.500000   
max    166.000000  301.000000  44039.000000  18355.000000  800.000000   

             Ave       Econ         SR          5         10  
count  79.000000  79.000000  79.000000  79.000000  79.000000  
mean   27.466456   2.806582  59.187342  16.354430   2.797468  
std     3.657561   0.351666   9.349337   9.642372   3.235935  
min    20.940000   1.980000  41.200000   3.000000   0.000000  
25%    24.425000   2.600000  53.300000   9.500000   1.000000  
50%    28.000000   2.820000  57.400000  14.000000   2.000000  
75%    29.870000   3.080000  63.950000  20.500000   3.500000  
max    34.790000   3.460000  91.900000  67.000000  22.000000

# Descriptive Statistics:


shows the details of all numerical attributes in the dataset:

1. count = number of observations
2. mean = average of all values
3. std = standard deviation
4. min = minimum value among all the observations
5. 25% = the value at 25th percentile  
6. 50% = median value 
7. 75% = the value at 75th percentile.
8. max = maximum value among all the observations. 


**we can see that 25%,50%,75% values or observation has less than or falls below that point**


**Observations:**


1. The average number of matches played by the bowlers is 80 and minimum is 37 matches and maximum is 166. 25% bowler played approximately or less than 60 matches. 50% bowler played around or less than 71 matches, 75% bowler played nearly or less than 93 matches.


2. The average number of innings played by the bowlers is 144 and minimum is 67 ,maximum is 301. 25% bowler played approximately or less than 110 innings. 50% bowler played around or less than 129 innings, 75% bowler played nearly or less than 169 innings.



3. The average number of runs conceded by the bowlers is 8595. While the highest runs is 18355, only 25%  bowler scored less than 6456  runs.




4. The average number of wickets taken by the bowlers is 317, with 75% of bowlers scoring less than 375.



5. The average number of 5 wickets taken by the bowlers is 16 while 75 % bowlers scored less than 20.

In [145]:
print(df.info())

# checking for missing values and data types of each column

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Player  79 non-null     object 
 1   Span    79 non-null     object 
 2   Mat     79 non-null     int64  
 3   Inns    79 non-null     int64  
 4   Balls   79 non-null     int64  
 5   Runs    79 non-null     int64  
 6   Wkts    79 non-null     int64  
 7   BBI     79 non-null     object 
 8   BBM     79 non-null     object 
 9   Ave     79 non-null     float64
 10  Econ    79 non-null     float64
 11  SR      79 non-null     float64
 12  5       79 non-null     int64  
 13  10      79 non-null     int64  
dtypes: float64(3), int64(7), object(4)
memory usage: 8.8+ KB
None


**Observations:**


* we can see that there is no missing values present in the dataset

* there is 4 object data type (player, span, BBI, BBM)

* there is 3 float data type ( Strike Rate, Economy Rate, Average) and others are integer value.

# Renaming the columns

In [146]:
df = df.rename(columns={'Mat':'Matches', 
                        'Inns':'Innings',
                        'Wkts':'Wickets',
                        'Ave': 'Bowling Average',
                        'Econ' : 'Economy Rate',
                        'SR' : 'Strike Rate',
                        'BBI' : 'Best Bowling in Innings',
                        'BBM' : 'Best Bowling in Match',
                        5: '5_wickets',
                        10:'10_wickets'})


df.head()

Player       Span  Matches  Innings  Balls   Runs  \
0  M Muralitharan (ICC/SL)  1992-2010      133      230  44039  18180   
1           SK Warne (AUS)  1992-2007      145      273  40705  17995   
2         A Kumble (INDIA)  1990-2008      132      236  40850  18355   
3        JM Anderson (ENG)  2003-2021      162      301  34791  16457   
4         GD McGrath (AUS)  1993-2007      124      243  29248  12186   

   Wickets Best Bowling in Innings Best Bowling in Match  Bowling Average  \
0      800     1951-09-01 00:00:00                16/220            22.72   
1      708     1971-08-01 00:00:00                12/128            25.41   
2      619     1974-10-01 00:00:00                14/149            29.65   
3      617     1942-07-01 00:00:00   1971-11-01 00:00:00            26.67   
4      563     2021-08-24 00:00:00   2021-10-27 00:00:00            21.64   

   Economy Rate  Strike Rate  5_wickets  10_wickets  
0          2.47         55.0         67          22  
1          2.65         57.4         37          10  
2          2.69         65.9         35           8  
3          2.83         56.3         30           3  
4          2.49         51.9         29           3

**Observations:**

Muralitharan was the best bowling in a match where he conceded 220 runs and number of taken wickets was 16. 

His average rate is 22.72 ,economy rate was 2.47 and number of 5 wicktes taken in a match was 67.

# Remove the columns BBI and BBM

we can see that the data in the column BBI and BBM was bugged and some of the entries got turned into datetime

In [147]:
df.drop(["Best Bowling in Match", "Best Bowling in Innings"], axis=1 , inplace= True )

display(df.head())

Player       Span  Matches  Innings  Balls   Runs  \
0  M Muralitharan (ICC/SL)  1992-2010      133      230  44039  18180   
1           SK Warne (AUS)  1992-2007      145      273  40705  17995   
2         A Kumble (INDIA)  1990-2008      132      236  40850  18355   
3        JM Anderson (ENG)  2003-2021      162      301  34791  16457   
4         GD McGrath (AUS)  1993-2007      124      243  29248  12186   

   Wickets  Bowling Average  Economy Rate  Strike Rate  5_wickets  10_wickets  
0      800            22.72          2.47         55.0         67          22  
1      708            25.41          2.65         57.4         37          10  
2      619            29.65          2.69         65.9         35           8  
3      617            26.67          2.83         56.3         30           3  
4      563            21.64          2.49         51.9         29           3

# Splitting Player and Country from 'Player' Column

in the player column ,there is information in the braket sign about the player   played for which country and ICC

In [148]:
df['Player'] = df['Player'].str.replace(")"," ")  
df_player = df['Player'].str.split("(", expand=True)
df = df.drop('Player', axis=1)  
df = pd.concat([df, df_player], axis=1) 
df = df.rename(columns={0: 'Player',1: 'Country'})  
column_sequence = ['Player' , 'Country' , 'Span' , 'Matches' , 'Innings', 'Balls', 'Runs' , 'Wickets', 'Bowling Average' , 'Economy Rate' , 'Strike Rate' , '5_wickets' , '10_wickets']

df = df[column_sequence]

display(df.head())

<ipython-input-148-399e9fb33449>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  df['Player'] = df['Player'].str.replace(")"," ")


Player  Country       Span  Matches  Innings  Balls   Runs  \
0  M Muralitharan   ICC/SL   1992-2010      133      230  44039  18180   
1        SK Warne      AUS   1992-2007      145      273  40705  17995   
2        A Kumble    INDIA   1990-2008      132      236  40850  18355   
3     JM Anderson      ENG   2003-2021      162      301  34791  16457   
4      GD McGrath      AUS   1993-2007      124      243  29248  12186   

   Wickets  Bowling Average  Economy Rate  Strike Rate  5_wickets  10_wickets  
0      800            22.72          2.47         55.0         67          22  
1      708            25.41          2.65         57.4         37          10  
2      619            29.65          2.69         65.9         35           8  
3      617            26.67          2.83         56.3         30           3  
4      563            21.64          2.49         51.9         29           3

# Finding how many players played for ICC

In [149]:
def ICC(x):
    if "ICC" in x:
        return "Yes"
    else:
        return "No"

In [150]:
df['Played for ICC'] = df['Country'].apply(ICC)

#display(df.head(10))

# removing "ICC/" from string of country column
df['Country'] = df['Country'].str.replace("ICC/","", regex=True)
display(df.head(5))

Player Country       Span  Matches  Innings  Balls   Runs  \
0  M Muralitharan      SL   1992-2010      133      230  44039  18180   
1        SK Warne     AUS   1992-2007      145      273  40705  17995   
2        A Kumble   INDIA   1990-2008      132      236  40850  18355   
3     JM Anderson     ENG   2003-2021      162      301  34791  16457   
4      GD McGrath     AUS   1993-2007      124      243  29248  12186   

   Wickets  Bowling Average  Economy Rate  Strike Rate  5_wickets  10_wickets  \
0      800            22.72          2.47         55.0         67          22   
1      708            25.41          2.65         57.4         37          10   
2      619            29.65          2.69         65.9         35           8   
3      617            26.67          2.83         56.3         30           3   
4      563            21.64          2.49         51.9         29           3   

  Played for ICC  
0            Yes  
1             No  
2             No  
3             No  
4             No

In [151]:
print (df['Played for ICC'].value_counts())

No     74
Yes     5
Name: Played for ICC, dtype: int64


**only 5 players played for ICC**

# Checking how many different countries are present in the dataset

In [152]:
print (df['Country'].value_counts())

AUS         18
ENG         13
INDIA       10
WI           9
SA           8
NZ           7
PAK          7
SL           3
ENG/ICC      2
ZIM          1
BDESH        1
Name: Country, dtype: int64


**Obsevations:**

**we can see that there is  18 Australian and 1 Bangladedshi bowlers present in the dataset**

# Splitting the span column

In [153]:
# splitting 'span' column
df_span = df['Span'].str.split("-", expand=True)

# concatenating the new dataframe with the main dataframe
df = pd.concat([df, df_span], axis=1)

# renaming new splitted column
df = df.rename(columns={0: "Starting Year",
                        1: "Ending Year"})

# removing 'span' column
df = df.drop("Span", axis=1)

# new column sequence as a list
column_sequence = ['Player', 'Country','Starting Year', 'Ending Year', 'Matches', 'Innings', 'Balls', 'Runs','Wickets' , 'Bowling Average','Economy Rate','Strike Rate', '5_wickets', '10_wickets' , 'Played for ICC']   

# implementing the new column sequence
df = df[column_sequence]


display(df.head(5))

Player Country Starting Year Ending Year  Matches  Innings  Balls  \
0  M Muralitharan      SL           1992        2010      133      230  44039   
1        SK Warne     AUS           1992        2007      145      273  40705   
2        A Kumble   INDIA           1990        2008      132      236  40850   
3     JM Anderson     ENG           2003        2021      162      301  34791   
4      GD McGrath     AUS           1993        2007      124      243  29248   

    Runs  Wickets  Bowling Average  Economy Rate  Strike Rate  5_wickets  \
0  18180      800            22.72          2.47         55.0         67   
1  17995      708            25.41          2.65         57.4         37   
2  18355      619            29.65          2.69         65.9         35   
3  16457      617            26.67          2.83         56.3         30   
4  12186      563            21.64          2.49         51.9         29   

   10_wickets Played for ICC  
0          22            Yes  
1          10             No  
2           8             No  
3           3             No  
4           3             No

In [154]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Player           79 non-null     object 
 1   Country          79 non-null     object 
 2   Starting Year    79 non-null     object 
 3   Ending Year      79 non-null     object 
 4   Matches          79 non-null     int64  
 5   Innings          79 non-null     int64  
 6   Balls            79 non-null     int64  
 7   Runs             79 non-null     int64  
 8   Wickets          79 non-null     int64  
 9   Bowling Average  79 non-null     float64
 10  Economy Rate     79 non-null     float64
 11  Strike Rate      79 non-null     float64
 12  5_wickets        79 non-null     int64  
 13  10_wickets       79 non-null     int64  
 14  Played for ICC   79 non-null     object 
dtypes: float64(3), int64(7), object(5)
memory usage: 9.4+ KB


# Changing Starting and Ending Year from object type to integer

In [155]:
#if the staring year and ending year were object type then we need to convert this to integer.
df['Ending Year']=pd.to_numeric(df['Ending Year'])
df['Starting Year']=pd.to_numeric(df['Starting Year'])

In [156]:
# total years of span of the players
df['Span']=df['Ending Year']-df['Starting Year']

# rearranging columns
column_sequence = ['Player', 'Country','Starting Year', 'Ending Year', 'Span' , 'Matches', 'Innings', 'Balls', 'Runs','Wickets' , 'Bowling Average','Economy Rate','Strike Rate', '5_wickets', '10_wickets' , 'Played for ICC']   
df = df[column_sequence] 
display(df.head(5))

Player Country  Starting Year  Ending Year  Span  Matches  \
0  M Muralitharan      SL            1992         2010    18      133   
1        SK Warne     AUS            1992         2007    15      145   
2        A Kumble   INDIA            1990         2008    18      132   
3     JM Anderson     ENG            2003         2021    18      162   
4      GD McGrath     AUS            1993         2007    14      124   

   Innings  Balls   Runs  Wickets  Bowling Average  Economy Rate  Strike Rate  \
0      230  44039  18180      800            22.72          2.47         55.0   
1      273  40705  17995      708            25.41          2.65         57.4   
2      236  40850  18355      619            29.65          2.69         65.9   
3      301  34791  16457      617            26.67          2.83         56.3   
4      243  29248  12186      563            21.64          2.49         51.9   

   5_wickets  10_wickets Played for ICC  
0         67          22            Yes  
1         37          10             No  
2         35           8             No  
3         30           3             No  
4         29           3             No

# Players with Longest to Shortest Span

In [157]:
display(df.sort_values(by='Span', ascending = False).head(1))

display(df.sort_values(by='Span', ascending = True).head(1))

Player Country  Starting Year  Ending Year  Span  Matches  Innings  \
21  Imran Khan     PAK            1971         1992    21       88      142   

    Balls  Runs  Wickets  Bowling Average  Economy Rate  Strike Rate  \
21  19458  8258      362            22.81          2.54         53.7   

    5_wickets  10_wickets Played for ICC  
21         23           6             No

Player Country  Starting Year  Ending Year  Span  Matches  Innings  \
44  GP Swann     ENG            2008         2013     5       60      109   

    Balls  Runs  Wickets  Bowling Average  Economy Rate  Strike Rate  \
44  15349  7642      255            29.96          2.98         60.1   

    5_wickets  10_wickets Played for ICC  
44         17           3             No

**Observations:**

* Imran Khan had longest carrer span. He had played for the longest period of time of 21 years.
* GP Swann Khan had shortest carrer span of 5 years

# Finding the player that had a lowest economy rate

In [158]:
display(df.sort_values(by='Economy Rate', ascending = True).head(5))

Player Country  Starting Year  Ending Year  Span  Matches  Innings  \
32      LR Gibbs      WI            1958         1976    18       79      148   
47      R Benaud     AUS            1952         1964    12       63      116   
35  DL Underwood     ENG            1966         1982    16       86      151   
39       BS Bedi   INDIA            1966         1979    13       67      118   
68   CV Grimmett     AUS            1925         1936    11       37       67   

    Balls  Runs  Wickets  Bowling Average  Economy Rate  Strike Rate  \
32  27115  8989      309            29.09          1.98         87.7   
47  19108  6704      248            27.03          2.10         77.0   
35  21862  7674      297            25.83          2.10         73.6   
39  21364  7637      266            28.71          2.14         80.3   
68  14513  5231      216            24.21          2.16         67.1   

    5_wickets  10_wickets Played for ICC  
32         18           2             No  
47         16           1             No  
35         17           6             No  
39         14           1             No  
68         21           7             No

**0bservations:**
* LR Gibbs had the lowest economy rate of 1.98

# Finding the player that had the lowest strike rate

In [159]:
display(df.sort_values(by='Strike Rate', ascending = True).head(5))

Player Country  Starting Year  Ending Year  Span  Matches  Innings  \
71      K Rabada      SA            2015         2021     6       47       86   
7       DW Steyn      SA            2004         2019    15       93      171   
20  Waqar Younis     PAK            1989         2003    14       87      154   
19   MD Marshall      WI            1978         1991    13       81      151   
25     AA Donald      SA            1992         2002    10       72      129   

    Balls   Runs  Wickets  Bowling Average  Economy Rate  Strike Rate  \
71   8785   4846      213            22.75          3.30         41.2   
7   18608  10077      439            22.95          3.24         42.3   
20  16224   8788      373            23.56          3.25         43.4   
19  17584   7876      376            20.94          2.68         46.7   
25  15519   7344      330            22.25          2.83         47.0   

    5_wickets  10_wickets Played for ICC  
71         10           4             No  
7          26           5             No  
20         22           5             No  
19         22           4             No  
25         20           3             No

**Observations:**
    
    K Rabada had the lowest strike rate of 41.2

# Finding the player that had the lowest bowling average

In [160]:
display(df.sort_values(by='Bowling Average', ascending = True).head(5))

Player Country  Starting Year  Ending Year  Span  Matches  Innings  \
19  MD Marshall      WI            1978         1991    13       81      151   
41     J Garner      WI            1977         1987    10       58      111   
15  CEL Ambrose      WI            1988         2000    12       98      179   
33   FS Trueman     ENG            1952         1965    13       67      127   
4    GD McGrath     AUS            1993         2007    14      124      243   

    Balls   Runs  Wickets  Bowling Average  Economy Rate  Strike Rate  \
19  17584   7876      376            20.94          2.68         46.7   
41  13169   5433      259            20.97          2.47         50.8   
15  22103   8501      405            20.99          2.30         54.5   
33  15178   6625      307            21.57          2.61         49.4   
4   29248  12186      563            21.64          2.49         51.9   

    5_wickets  10_wickets Played for ICC  
19         22           4             No  
41          7           0             No  
15         22           3             No  
33         17           3             No  
4          29           3             No

**Observations:**
     
    MD Marshall had the lowest bowling average of 20.94